server.py

```diff
@@ -1,69 +1,134 @@
 import bottle
 import os, sys, datetime
 import string, random
-
-from collections import defaultdict, namedtuple
-import shelve
+from peewee import *
 
 path = os.path.abspath(__file__)
 dir_path = os.path.dirname(path)
 app = bottle.Bottle()
 
-database_path = "submission_record.db"
-user_db = "user_record.db"
-sessions_db = "session_record.db"
-questions = {}
-contests = {}
+DATABASE_NAME = "data.db"
 question_dir = "files/questions"
 
-Question = namedtuple("Question", "output statement")
-Submission = namedtuple("Submission", "question time output is_correct contest")
-Contest = namedtuple("Contest", "description questions start_time end_time")
-User = namedtuple("User", "password")
+db = SqliteDatabase(DATABASE_NAME)
+
+
+class User(Model):
+    username = CharField(unique=True)
+    password = CharField()
+
+    class Meta:
+        database = db
+
+
+class Session(Model):
+    def random_token():
+        return "".join([random.choice(string.ascii_letters) for _ in range(20)])
+
+    token = CharField(unique=True, default=random_token)
+    user = ForeignKeyField(User)
+
+    class Meta:
+        database = db
+
+
+class Contest(Model):
+    code = CharField(unique=True)
+    description = CharField()
+    start_time = DateTimeField()
+    end_time = DateTimeField()
+
+    class Meta:
+        database = db
+
+
+class Question(Model):
+    q_no = IntegerField(unique=True)
+    author = ForeignKeyField(User)
+
+    class Meta:
+        database = db
+
+
+class ContestProblems(Model):
+    contest = ForeignKeyField(Contest, backref="questions")
+    question = ForeignKeyField(Question)
+
+    class Meta:
+        database = db
+        indexes = ((("contest", "question"), True),)
+
+
+class Submission(Model):
+    user = ForeignKeyField(User)
+    time = DateTimeField()
+    question = ForeignKeyField(Question)
```

Comment:
since a `ContestProblem` refers to both a contest and a problem let's have a foreign key to that?